### ISSUE TITLE
[BUG] POSTs to `/v3/cloudcredentials` succeed but fail

### ISSUE THREAD

**Issue by richard-cox** [rancher/rancher#41745]
**Rancher Server Setup**
- Rancher version: 	v2.7-f0d4078f4fa75b303ce276c3f4b12fc465b1ba2a-head
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: 1.25
- Cluster Type (Local/Downstream): local
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - Admin

**Describe the bug**
- POST to `/v3/cloudcredenitals` with the below content returns 201 succeeds but does returns `null` and no credentials created
- Bug 1 - Response should not be 201 if the call fails
- Bug 2 - Response should not fail

Fails - Note - The Cloud CA node driver's cloud config schema does not contain any `resourceFields`
```
{"type":"cloudcredential","metadata":{"namespace":"fleet-default","annotations":{"provisioning.cattle.io/driver":"cloudca"},"generateName":"cloud-credential-"},"cloudcacredentialConfig":{"apiKey":"asdsad"},"name":"asdsadasdas"}
```

Succeeds - Note - The OTC node driver's cloud config schema does contain any `resourceFields`
```
{"type":"cloudcredential","metadata":{"namespace":"fleet-default","annotations":{"provisioning.cattle.io/driver":"otc"},"generateName":"cloud-credential-"},"otccredentialConfig":{"accessKeySecret":"otc"},"name":"otc"}
```

**To Reproduce**
In the UI...
- Enable Cloud CA node driver (don't refresh)
- Navigate to Cluster Management --> Clusters --> Create button --> RKE2 is toggled on --> Cloud.ca
- Enter a name and apiKey
- click Continue

**Result**
- 201 response, user should then proceed to rke2 cluster create screen... however as no cloud cred resource returned in create request we're in a weird world

**Expected Result**
- If request contains junk the response should be 4XX 
- If request is valid response should be 201 AND contain the new resourcereturned 

**Screenshots**

**Additional context**

![243418803-4ee989b6-b7fa-4bc8-a223-bfa1a1240c29](https://github.com/rancher/rancher/assets/18697775/472339ff-5831-4804-9c32-5ec43424603e)

**Comment by gaktive** [comment]
As noted, this blocks https://github.com/rancher/dashboard/issues/8523 for UI. Pushing that out of Q2 for now as time is tight.

**Comment by gaktive** [comment]
Internal reference: SURE-6517

**Comment by gaktive** [comment]
Jira item now deleted but this ticket remains in limbo.

@jakefhyde since you brought up https://github.com/rancher/dashboard/issues/8523 earlier, are you blocked by this? If so, we can work with @samjustus on prioritization for Q1.

**Comment by MbolotSuse** [comment]
> @MbolotSuse , thank you for doing thorough research into this issue. While it makes sense that schema is required for properly processing this request, it doesn't justify getting 201 (CREATED) with null response - a more fitting HTTP response code would be somethis in 4xx (or possibly 5xx) range. It should probably be spawned into separate issue to be owned by the frameworks team, something original reported alluded to:

> > Bug 1 - Response should not be 201 if the call fails
> > Bug 2 - Response should not fail

> If you agree, could you / @samjustus / @cbron create this issue?

@snasovich Thanks for your detailed reply. Unfortunately, I do not agree that ownership of the 201 falls within the frameworks team. As briefly alluded to in my original comment, I believe that the issue causing the 201 is that a cloud-credential specific customization is returning a nil result and no error in this case. That code can be found [here](https://github.com/rancher/rancher/blob/c5e8992b2337472a7373859ef20c3ecd5c582a57/pkg/api/norman/customization/cred/store.go#L32). Since that function is specific to cloud credentials, it won't be owned by the Frameworks team (which does not own resource specific customizations). 

In short, Norman is returning a 201 and empty reply because the cloud credential endpoint has been misconfigured to do so. Since that misconfiguration is specific to that endpoint, it falls within the owning teams' area of responsibility.

If upon further investigation you find that I am incorrect about the source of the 201, then I would agree that the frameworks team should analyze further.

As far as the rest of your reply goes, I agree with the current analysis of the issue and recommendation that we pursue an enhancement to not require a hardcoded change in rancher to support things like this, though I'm not an expert in this area, so that's just an opinion on my end.

**Comment by cbron** [comment]
I'd agree that if no error is returned then 2XX seems in line with go standards, though the null response is weird given situation. IMO the frameworks aren't advanced enough to gracefully handle all cascading error situations like this. If provisioning team can take a look, you can let us know if anything else is needed, but I imagine fixing that endpoint should solve everything.

**Comment by palvarez89** [comment]
> However, following some investigations I believe there is a **reasonable workaround** that Rancher admin can apply once and it "sticks" even through Rancher restarts / driver updates.
> 
> The workaround is to simply update `annotations` on corresponding `nodedrivers.management.cattle.io` object to set `privateCredentialFields` / `publicCredentialFields` to set as comma-separated list of fields that should be public/private fields on Cloud Credential for that driver. This change can be done via `kubectl` (recommended) 

Thank you for this workaround, it helped us to test our own node driver doing RKE2 cluster deployments. Is there a way to  change these `privateCredentialFields` / `publicCredentialFields`  ?  I was only able to change those by re-installing rancher again.

**Comment by snasovich** [comment]
> > However, following some investigations I believe there is a **reasonable workaround** that Rancher admin can apply once and it "sticks" even through Rancher restarts / driver updates.
> > The workaround is to simply update `annotations` on corresponding `nodedrivers.management.cattle.io` object to set `privateCredentialFields` / `publicCredentialFields` to set as comma-separated list of fields that should be public/private fields on Cloud Credential for that driver. This change can be done via `kubectl` (recommended)
> 
> Thank you for this workaround, it helped us to test our own node driver doing RKE2 cluster deployments. Is there a way to change these `privateCredentialFields` / `publicCredentialFields` ? I was only able to change those by re-installing rancher again.

@palvarez89 , I've just tried updating `publicCredentialFields` annotation on the `nodeDriver` object to add extra field to it and it immediately showed in `resourceFields` of the `v3/dynamicSchemas/(drivername)credentialconfig` link and after refreshing UI I see this new field appearing on form for creating cloud credentials for this node driver.

Could you please provide more information on what changes require Rancher reinstall for you to take effect?

**Comment by snasovich** [comment]
@cbron @MbolotSuse , just to close the loop on the above conversations, I agree that the [transformer](https://github.com/rancher/rancher/blob/c5e8992b2337472a7373859ef20c3ecd5c582a57/pkg/api/norman/customization/cred/store.go#L24) should not return `nil, nil` as it's essentially stating “there was no error but there is no object available” which doesn’t make sense. I don't believe there is a way in Go to prevent `nil, nil` response so it's up for caller func to handle such a response.
This is then [handled](https://github.com/rancher/norman/blob/5de27f66385d72084a81fb3bf972dd89dd2fff77/api/handler/create.go#L23-L29) by Norman's API handler which produces that "201 - null" response.
We're splitting hairs here but IMO in a perfect world the system that depends on subsystems outside of its control should be resilient to potentially invalid responses from those subsystems and still produce valid responses. In a perfect world response 500 with error like “object not created, no further details are available” makes more sense than 201 with null content.

I'll leave it up to your team to see if it makes sense to add more resiliency to the code I link and for this issue we'll focus on ensuring Transformer doesn't return `nil, nil` for custom node drivers. Like mentioned in previous response, it's going to be an enhancement that, once implemented, will need to be adopted by individual node driver implementations to support:
> Now, focusing on "Bug 2", adding `cloudca` to `DriverData` map [here](https://github.com/rancher/rancher/blob/c5e8992b2337472a7373859ef20c3ecd5c582a57/pkg/data/management/machinedriver_data.go#L38) will certainly address the immediate issue for Cloud.ca driver but we need a more generic solution to allow node driver binary itself to declare `publicCredentialFields` / `privateCredentialFields` (and while at it, `defaults` as defined [here](https://github.com/rancher/rancher/blob/c5e8992b2337472a7373859ef20c3ecd5c582a57/pkg/data/management/machinedriver_data.go#L58)) as having these needed to be hardcoded in Rancher goes against "plugin" model of node drivers.
This is an **enhancement** to be designed and implemented.

**Comment by palvarez89** [comment]
> @palvarez89 , I've just tried updating `publicCredentialFields` annotation on the `nodeDriver` object to add extra field to it and it immediately showed in `resourceFields` of the `v3/dynamicSchemas/(drivername)credentialconfig` link and after refreshing UI I see this new field appearing on form for creating cloud credentials for this node driver.
> 
> Could you please provide more information on what changes require Rancher reinstall for you to take effect?

I've just tried again, and yes, it looks like the UI gets updated immediately. I'm now having problems to set fields that contain a hyphen in them, like for example `image-id`. Is there a way to escape them without changing the driver?
